core/src/main/java/org/apache/iceberg/rest/auth/AuthSessionCache.java

```diff
@@ -0,0 +1,158 @@
+ * under the License.
+ */
+package org.apache.iceberg.rest.auth;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import java.time.Duration;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ThreadFactory;
+import java.util.concurrent.TimeUnit;
+import java.util.function.Function;
+import java.util.function.LongSupplier;
+import javax.annotation.Nullable;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.iceberg.relocated.com.google.common.util.concurrent.Uninterruptibles;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/** A cache for {@link AuthSession} instances. */
+public class AuthSessionCache implements AutoCloseable {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AuthSessionCache.class);
+
+  private final Duration sessionTimeout;
+  private final Executor executor;
+  private final LongSupplier nanoTimeSupplier;
+
+  private volatile Cache<String, AuthSession> sessionCache;
+
+  /**
+   * Creates a new cache with the given session timeout, and with default executor and nano time
+   * supplier for eviction tasks.
+   *
+   * @param name a distinctive name for the cache.
+   * @param sessionTimeout the session timeout. Sessions will become eligible for eviction after
+   *     this duration of inactivity.
+   */
+  public AuthSessionCache(String name, Duration sessionTimeout) {
+    this(name, sessionTimeout, null, null);
+  }
+
+  /**
+   * Creates a new cache with the given session timeout, executor, and nano time supplier. This
+   * method is useful for testing mostly.
+   *
+   * @param name a distinctive name for the cache.
+   * @param sessionTimeout the session timeout. Sessions will become eligible for eviction after
+   *     this duration of inactivity.
+   * @param executor the executor to use for eviction tasks; if null, the cache will create a
+   *     default executor. The executor will be closed when this cache is closed.
+   * @param nanoTimeSupplier the supplier for nano time; if null, the cache will use {@link
+   *     System#nanoTime()}.
+   */
+  AuthSessionCache(
+      String name,
+      Duration sessionTimeout,
+      @Nullable Executor executor,
+      @Nullable LongSupplier nanoTimeSupplier) {
+    this.sessionTimeout = sessionTimeout;
+    this.executor = executor == null ? createExecutor(name) : executor;
+    this.nanoTimeSupplier = nanoTimeSupplier;
+  }
+
+  private static Executor createExecutor(String name) {
+    ThreadFactory threadFactory =
+        new ThreadFactoryBuilder()
+            .setNameFormat(name + "-auth-session-evict-%d")
+            .setDaemon(true)
+            .build();
+    return Executors.newCachedThreadPool(threadFactory);
```

Comment:
I think this is the best executor for this kind of task, as we don't want a core thread kept alive if there isn't anything to do.

Reply:
We should be using a dedicated pool from our `ThreadPools` utility.  I would suggest: `ThreadPools::newExitingWorkerPool(String namePrefix, int poolSize)` like we had previously.

Reply:
OK, switched to `ThreadPools.newExitingWorkerPool`.